### ISSUE TITLE
Assembly-level Action Attribute doesn't run actions on fixtures

### ISSUE THREAD

**Issue by oznetmaster** [nunit/nunit#1770]
Is there any way to easily write a global SetUp or TearDown method?

This would be like a SeUp or TearDown that would be run once for each TestFixture in the test suite.

If not, it seems like it would be very useful to have one.

**Comment by CharliePoole** [comment]
You'd have to take a closer look, but I think you can create an ActionAttribute that does that.

**Comment by oznetmaster** [comment]
Similarly, how about a really global SetUp/TearDown that encompasses the entire test suite, and is run once when the test suite is started, and once after it completes?

The problem is that in my CF environment, once the test assembly is loaded, it is there forever until the entire test environment is stopped and reloaded, which can take about 1 minute.

So if multiple tests are run against the same test assembly, there is sometimes a need to reset certain things before the run.

**Comment by CharliePoole** [comment]
Depends what you mean. We have OneTimeSetUp using SetUpFixtures, which can encompass any namespace or the entire assembly. But if a test run involves multiple assemblies, there is no run-level OneTimeSetUp. Creating one  would be problematic, since test assemblies may run in multiple processes and/or AppDomains.

**Comment by oznetmaster** [comment]
It is unclear from the documentation if attaching an ActionAttribute at the assembly level will only run the methods once per assembly, or will still run them for each TestFixture in the assembly.

If the first, then that is the solution.  If the later, then a new ActionTarget seems useful, Assembly.

**Comment by oznetmaster** [comment]
SetUpFixture??  I cannot find a reference to that in the documentation.

**Comment by oznetmaster** [comment]
My CF environment only has a single AppDomain and a singe process, so those are not issues.

**Comment by oznetmaster** [comment]
Found SetUpFixture.  Did nor realize it was an attribute.

**Comment by oznetmaster** [comment]
That actually seems to meet the needs I currently have. :)

**Comment by CharliePoole** [comment]
O

**Comment by CharliePoole** [comment]
Well good. We aim to please!

Close this?

**Comment by Sebazzz** [comment]
> This would be like a SeUp or TearDown that would be run once for each TestFixture in the test suite.

According to the docs, an `ITestAction` should be used for that. Example:

``` C#
[AttributeUsage(AttributeTargets.Assembly)]
public sealed class MyActionAttribute : Attribute, ITestAction {
    // [removed for brevity]
    public void BeforeTest(ITest test) {
        this._logger.Log((test.ClassName ?? "_") + "." + (test.MethodName ?? "_"));
    }
    public void AfterTest(ITest test) {
        this._logger.Log((test.ClassName ?? "_") + "." + (test.MethodName ?? "_"));
    }
    public ActionTargets Targets => ActionTargets.Suite | ActionTargets.Test;
}
```

According to the docs it should run on test suites and test methods. But, it doesn't run on TestFixture. Is that correct behaviour @CharliePoole? Since a Test Fixture "is" a test suite, shouldn't it run on that as well? As a kind-of globally applied OneTimeSetUp and OneTimeTearDown per test fixture? If not, what would be the alternatives?

If the behaviour is correct, I propose to explicitly state this behaviour in the docs.

**Comment by CharliePoole** [comment]
@Sebazzz Generally, when we talk about a TestSuite in NUnit, we include any composed test that includes child tests under it. That includes a TestFixture. Could you give me a reference to the documentation that seems unclear?

If the example you posted does __not__ run before each fixture as well as before each test case, I think that may be a bug.

**Comment by CharliePoole** [comment]
Copy of comment by @Sebazzz posted to a new issue by mistake:

-----
As discussed in [#1770](https://github.com/nunit/nunit/issues/1770#issuecomment-247627302), when an ITestAction is placed on the assembly level it does not run for test suites within the assembly. 

Please review an reproduction with [this gist](https://gist.github.com/Sebazzz/5c04b92f566bc8e9c6397b2f10d4f191). Compile by placing NUnit 3 console runner and NUnit 3 assembly into same directory and running 'run.bat'.

**Actual output:**
```
Before: C:\temp\nunittestaction\Test.dll
Before: TestActions.MyTestFixture.Test1
MyTestFixture.Test1
After: TestActions.MyTestFixture.Test1
Before: TestActions.MyTestFixture.Test2
MyTestFixture.Test2
After: TestActions.MyTestFixture.Test2
After: C:\temp\nunittestaction\Test.dll
```

**Expected output:**
```
Before: C:\temp\nunittestaction\Test.dll
Before: TestActions [note: since an namespace is an test suite too]
Before: TestActions.MyTestFixture
Before: TestActions.MyTestFixture.Test1
MyTestFixture.Test1
After: TestActions.MyTestFixture.Test1
Before: TestActions.MyTestFixture.Test2
MyTestFixture.Test2
After: TestActions.MyTestFixture.Test2
After: TestActions.MyTestFixture
After: TestActions
After: C:\temp\nunittestaction\Test.dll
```

Reproduces with NUnit 3.4.1 (current version).

**Comment by CharliePoole** [comment]
@Sebazzz Could you add the info about which page in the docs you are referring to? I'd like to fix that if it's incorrect.

**Comment by Sebazzz** [comment]
Hi @CharliePoole, check the page on [action attributes](https://github.com/nunit/docs/wiki/Action-Attributes#AssemblyAttachedAction), chapter about Assembly Attached Action.

**Comment by CharliePoole** [comment]
That doc was originally written by the person who contributed the feature and seems a little ambiguous as to what it means by Suite. Internally, NUnit considers any aggregation of tests a suite, including an assembly, namespace, test fixture, parameterized method, etc. OTOH, his implementation does not deal with all suites suites, so he probably did not intend this to apply to suites that are not represented by an object of some kind (i.e. class or method).

I think we should roll the topic back to figuring out what's useful and why. Then we can make a plan to bring the docs and implementation into synch by changing one or the other - or both. Do you have a use case for wanting namespaces to be included? 

One thing to check: does it ignore all namespace suites or does the action apply to those with a SetUpFixture attached? And if so, would it run twice if there are two SetUpFixtures?

**Comment by Sebazzz** [comment]
No, I don't have an use case for running it on namespaces, but I do expect the methods to be run for fixtures themselves, which currently doesn't happen.

Test actions on the assembly are not run for any test suite - including testfixtures - regardless if any SetUpFixture is used. Only for each test method test actions are run.

**Comment by CharliePoole** [comment]
@sebazz You were no doubt correct to create a new issue in the first place, but now that I've put us back on this one, I'm changing the name of the issue to reflect what we're talking about. :smile: 

I wasn't able to access your gist from the cafe where I was working earlier. I can get it now and I'll give it a try before commenting further.

**Comment by rprouse** [comment]
There has been no discussion on this since 2016, so I am going to close. It can be reopened if someone wants to move it forward.
